hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py

```diff
@@ -0,0 +1,241 @@
+from enum import Enum
+import logging
+import math
+import numpy as np
+
+from health_cpath.preprocessing.loading import WSIBackend
+from health_cpath.utils.naming import SlideKey
+from monai.data.wsi_reader import WSIReader
+from monai.transforms import MapTransform
+from openslide import OpenSlide
+from pathlib import Path
+from tifffile.tifffile import TiffWriter, PHOTOMETRIC, COMPRESSION
+from typing import Any, Dict, List, Optional, Tuple
+
+
+AMPERSAND = "&"
+UNDERSCORE = "_"
+
+
+class WSIFormat(str, Enum):
+    """The format of the wsi file."""
+    NDPI = "ndpi"
+    SVS = "svs"
+    TIF = "tif"
+    TIFF = "tiff"
+
+
+class ConvertWSIToTiffd(MapTransform):
+    """Converts a wsi file to a tiff file. The tiff file is saved in the output_folder with the same name as the src
+    file but with the tiff extension. Ampersands are replaced by the replace_ampersand_by string. The tiff file
+    contains the image data at the target magnifications. If target_magnifications is None, the tiff file contains the
+    image data at all magnifications. If add_lowest_magnification is True, the tiff file also contains the image data
+    at the lowest magnification. The tiff file is saved with the compression specified by the compression parameter with
+    a fixed tile size.
+    """
+    OBJECTIVE_POWER_KEY = "openslide.objective-power"
+    RESOLUTION_UNIT_KEY = "tiff.ResolutionUnit"
+    RESOLUTION_UNIT = "centimeter"
+    SOFTWARE = "tifffile"
+
+    def __init__(
+        self,
+        output_folder: Path,
+        image_key: str = SlideKey.IMAGE,
+        src_format: WSIFormat = WSIFormat.NDPI,
+        target_magnifications: Optional[List[float]] = [10.],
+        add_lowest_magnification: bool = False,
+        base_objective_power: Optional[float] = None,
+        replace_ampersand_by: str = UNDERSCORE,
+        compression: COMPRESSION = COMPRESSION.ADOBE_DEFLATE,
+        tile_size: int = 512,
+        backend: WSIBackend = WSIBackend.OPENSLIDE,
+    ) -> None:
+        """
+        :param output_folder: The directory where the tiff file will be saved.
+        :param image_key: The key of the image in the data dictionary, defaults to SlideKey.IMAGE
+        :param src_format: The format of the src file, defaults to WSIFormat.NDPI
+        :param target_magnifications: The target magnifications e.g. [10., 20.], defaults to [10.]. If
+        target_magnifications is None, the tiff file will contain the image data at all magnifications.
```

Comment:
Here and across the whole PR: when doing documentation `:param` that span multiple lines, the second line needs to be indented.